components/merkledb/src/backends/rocksdb.rs

```diff
@@ -181,7 +202,14 @@ impl RocksDB {
             // SAFETY:
             // The snapshot carries an `Arc` to the database to make sure that database
             // is not dropped before the snapshot.
-            snapshot: unsafe { mem::transmute(self.db.snapshot()) },
+            snapshot: unsafe {
+                mem::transmute(
```

Comment:
Is `mem::transmute` on a reference obtained from lock legal?

Reply:
Good question 👍🏻 But I haven't met any problem or test fail. But I'll investigate this case much deeper.

Reply:
Logically, it looks unsafe: we forget a read lock on data, so a following write may invalidate some pointers inside `Snapshot`. At least this is what `fn snapshot(&self) -> Snapshot<'_>` signature entails from the Rust perspective; it means that `Snapshot` may contain refs to the database, which may be invalidated by DB mutation. I'm not sure it *does* entail that from the RocksDB perspective, since snapshots are explicitly supposed to not get invalidated by the writes.